drivers/adodb-pdo_firebird.inc.php

```diff
@@ -0,0 +1,447 @@
+				$fld->sub_type = $rs->fields[5];
+			} else {
+				$fld->sub_type = null;
+			}
+			if ($ADODB_FETCH_MODE == ADODB_FETCH_NUM) {
+				$retarr[] = $fld;
+			} else {
+				$retarr[strtoupper($fld->name)] = $fld;
+			}
+
+			$rs->MoveNext();
+		}
+		$rs->Close();
+		if (empty($retarr)) {
+			return false;
+		} else {
+			return $retarr;
+		}
+	}
+
+	public function metaIndexes($table, $primary = false, $owner = false)
+	{
+		// save old fetch mode
+		global $ADODB_FETCH_MODE;
+		$save = $ADODB_FETCH_MODE;
+		$ADODB_FETCH_MODE = ADODB_FETCH_NUM;
+		if ($this->fetchMode !== false) {
+			$savem = $this->SetFetchMode(false);
+		}
+		$table = strtoupper($table);
+		$sql = "SELECT * FROM RDB\$INDICES WHERE RDB\$RELATION_NAME = '" . $table . "'";
+		if (!$primary) {
+			$sql .= " AND RDB\$INDEX_NAME NOT LIKE 'RDB\$%'";
+		} else {
+			$sql .= " AND RDB\$INDEX_NAME NOT LIKE 'RDB\$FOREIGN%'";
+		}
+
+		// get index details
+		$rs = $this->Execute($sql);
+		if (!is_object($rs)) {
+			// restore fetchmode
+			if (isset($savem)) {
+				$this->SetFetchMode($savem);
+			}
+			$ADODB_FETCH_MODE = $save;
+			return false;
+		}
+
+		$indexes = array();
+		while ($row = $rs->FetchRow()) {
+			$index = $row[0];
+			if (!isset($indexes[$index])) {
+				if (is_null($row[3])) {
+					$row[3] = 0;
+				}
+				$indexes[$index] = array(
+					'unique' => ($row[3] == 1),
+					'columns' => array()
+				);
+			}
+			$sql = "SELECT * FROM RDB\$INDEX_SEGMENTS WHERE RDB\$INDEX_NAME = '" . $index . "' ORDER BY RDB\$FIELD_POSITION ASC";
+			$rs1 = $this->Execute($sql);
+			while ($row1 = $rs1->FetchRow()) {
+				$indexes[$index]['columns'][$row1[2]] = $row1[1];
+			}
+		}
+		// restore fetchmode
+		if (isset($savem)) {
+			$this->SetFetchMode($savem);
+		}
+		$ADODB_FETCH_MODE = $save;
+
+		return $indexes;
+	}
+
+	public function metaPrimaryKeys($table, $owner_notused = false, $internalKey = false)
+	{
+		if ($internalKey) {
+			return array('RDB$DB_KEY');
+		}
+
+		$table = strtoupper($table);
+
+		$sql = 'SELECT S.RDB$FIELD_NAME AFIELDNAME
+	FROM RDB$INDICES I JOIN RDB$INDEX_SEGMENTS S ON I.RDB$INDEX_NAME=S.RDB$INDEX_NAME
+	WHERE I.RDB$RELATION_NAME=\'' . $table . '\' and I.RDB$INDEX_NAME like \'RDB$PRIMARY%\'
+	ORDER BY I.RDB$INDEX_NAME,S.RDB$FIELD_POSITION';
+
+		$a = $this->GetCol($sql, false, true);
+		if ($a && sizeof($a) > 0) {
+			return $a;
+		}
+		return false;
+	}
+
+	public function createSequence($seqname = 'adodbseq', $startID = 1)
+	{
+		$ok = $this->execute("CREATE SEQUENCE $seqname");
+		if (!$ok) {
+			return false;
+		}
+
+		return $this->execute("ALTER SEQUENCE $seqname RESTART WITH " . ($startID - 1));
+	}
+
+	public function dropSequence($seqname = 'adodbseq')
+	{
+		$seqname = strtoupper($seqname);
+		return $this->Execute("DROP SEQUENCE $seqname");
+	}
+
+
+	public function _affectedrows()
+	{
+		return fbird_affected_rows($this->_transactionID ? $this->_transactionID : $this->_connectionID);
```

Comment:
@mnewnham Is this valid ? 

Function is not documented (there is [ibase_affected_rows()](https://www.php.net/manual/en/function.ibase-affected-rows.php))
Shouldn't we simply rely on the parent's ADODB_pdo::_affectedrows() ?

Reply:
@mnewnham I merged this, but I realize now that you never answered the above question...

Reply:
[fbird_affected_rows ](https://www.php.net/manual/en/function.fbird-affected-rows.php )appears to exist now